File: week2/chanki/백준 11286. 절댓값 힙.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Comparator;
import java.util.PriorityQueue;

public class B11286 {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        // 문제의 요구사항에 맞춰 힙 선언
        // 절댓값이 같으면 원본 값으로 비교
        PriorityQueue<int[]> pq = new PriorityQueue<>(
            Comparator.<int[]>comparingInt(o -> o[0])
                .thenComparing(o -> o[1])
        );
        StringBuilder sb = new StringBuilder();
        Integer inputCount = Integer.parseInt(br.readLine());
        for (int i = 0; i < inputCount; i++) {
            Integer input = Integer.parseInt(br.readLine());
            // 잘댓값과 원본 함께 저장
            if (input != 0) {
                pq.add(new int[]{Math.abs(input), input});
                continue;
            }

            if (pq.isEmpty()) {
                sb.append("0").append("\n");
                continue;
            }

            sb.append(pq.poll()[1]).append("\n");
        }

        System.out.println(sb);
    }
}

File: week2/chanki/백준 11286. 절댓값 힙.kt
import java.util.PriorityQueue
import java.util.Scanner
import kotlin.math.absoluteValue

fun main() = with(Scanner(System.`in`)) {
    // 문제의 요구사항에 맞춰 최소힙 선언
    // 절댓값이 같으면 원본을 비교
    val pq = PriorityQueue(compareBy<IntArray> { it[0] }.thenComparing { it -> it[1] })
    val inputCount = readlnOrNull()!!.toInt()
    val sb = StringBuilder()
    for (i in 1..inputCount) {
        val input = readlnOrNull()!!.toInt()
        // 조건에 맞춰 분기문
        when {
            input == 0 -> when (pq.isEmpty()) {
                true -> sb.append("0").append("\n")
                false -> sb.append(pq.poll()[1]).append("\n")
            }

            // 절댓값과 원본을 함께 저장
            else -> pq.add(intArrayOf(input.absoluteValue, input))
        }
    }

    println(sb)
}
